internal/olm/operator/registry/index_image.go

```diff
@@ -102,36 +87,58 @@ func (c IndexImageCatalogCreator) getDBPath(ctx context.Context) (string, error)
 	return defaultDBPath, nil
 }
 
-func (c IndexImageCatalogCreator) createRegistryPod(ctx context.Context, dbPath string) (*index.RegistryPod, error) {
-	// Create registry pod, assigning its owner as the catalog source
-	registryPod, err := index.NewRegistryPod(c.cfg.Client, dbPath, c.BundleImage, c.cfg.Namespace)
+func (c IndexImageCatalogCreator) createRegistryPod(ctx context.Context, dbPath string, cs *v1alpha1.CatalogSource) (*corev1.Pod, error) {
+	// Initialize registry pod
+	registryPod, err := index.NewRegistryPod(c.cfg, dbPath, c.BundleImage)
 	if err != nil {
-		return nil, fmt.Errorf("error in initializing registry pod")
+		return nil, fmt.Errorf("error initializing registry pod: %v", err)
 	}
 
-	if err = registryPod.Create(ctx); err != nil {
-		return nil, fmt.Errorf("error in creating registry pod")
+	var pod *corev1.Pod
+	// Create registry pod
+	if pod, err = registryPod.Create(ctx, cs); err != nil {
+		return nil, fmt.Errorf("error creating registry pod: %v", err)
 	}
 
-	return registryPod, nil
+	return pod, nil
 }
 
-func (c IndexImageCatalogCreator) updateCatalogSource(podAddr string, cs *v1alpha1.CatalogSource) error {
-	// Update catalog source with source type as grpc and address to point to the pod IP
-	cs.Spec.SourceType = v1alpha1.SourceTypeGrpc
-	cs.Spec.Address = index.GetRegistryPodHost(podAddr)
-
-	// Update catalog source with annotations for index image,
-	// injected bundle, and registry add mode
+func (c IndexImageCatalogCreator) updateCatalogSource(ctx context.Context, podAddr string, cs *v1alpha1.CatalogSource) error {
+	// JSON marshal injected bundles
 	injectedBundlesJSON, err := json.Marshal(c.InjectBundles)
 	if err != nil {
-		return fmt.Errorf("error in json marshal injected bundles: %v", err)
+		return fmt.Errorf("error marshaling injected bundles: %v", err)
+	}
+
+	// Get catalog source key
+	catsrcKey := types.NamespacedName{
+		Namespace: cs.GetNamespace(),
+		Name:      cs.GetName(),
 	}
-	cs.ObjectMeta.Annotations = map[string]string{
+
+	// Annotations for catalog source
+	annotationMapping := map[string]string{
 		"operators.operatorframework.io/index-image":        c.IndexImage,
 		"operators.operatorframework.io/inject-bundle-mode": c.InjectBundleMode,
 		"operators.operatorframework.io/injected-bundles":   string(injectedBundlesJSON),
 	}
+	// Update catalog source with source type as grpc and address as the pod IP,
+	// and annotations for index image, injected bundles, and registry bundle add mode
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := c.cfg.Client.Get(ctx, catsrcKey, cs); err != nil {
+			return fmt.Errorf("error getting catalog source: %v", err)
+		}
+		cs.Spec.Address = index.GetRegistryPodHost(podAddr)
+		cs.Spec.SourceType = v1alpha1.SourceTypeGrpc
+		cs.ObjectMeta.Annotations = annotationMapping
```

Comment:
You don't want to overwrite all annotations, just those with matching keys:
```suggestion
		annotations := cs.GetAnnotations()
		if annotations == nil {
			annotations = make(map[string]string, len(annotationMapping))
		}
		for k, v := range annotationMapping {
			annotations[k] = v
		}
		cs.SetAnnotations(annotations)
```